### PR TITLE
Implementation of ndimage filters [Continued #3184]

### PR DESCRIPTION
**Pull request by mnicely**
This PR applies @emcastillo's and @grlee77's suggestions to #3184. I believe I have successfully merged it with master and it passes all tests.

I'm still waiting for guidance on https://github.com/cupy/cupy/pull/3184#pullrequestreview-379274429

### CODE REVIEW

**Comment by emcastillo** [comment]
Thanks for all the changes and the fast reaction. 
I will try to send a review asap :)

**Comment by mnicely** [comment]
### Update
After further inspection, I believe the tests are actually being skipped, not passed.

File: cupyx/scipy/ndimage/filters.py [hunk 1]
```diff
@@ -78,39 +172,105 @@ def _get_output(output, input, shape=None):
...
+
+
+def _check_mode(mode):
     if mode not in ('reflect', 'constant', 'nearest', 'mirror', 'wrap'):
         msg = 'boundary mode not supported (actual: {}).'.format(mode)
         raise RuntimeError(msg)
+    return mode
+
+
+# def _check_axis(axis, ndim):
```
Comment (emcastillo) [inline cupyx/scipy/ndimage/filters.py:200]: Can you remove this please? 🙂

File: cupyx/scipy/ndimage/filters.py [hunk 2]
```diff
@@ -64,6 +65,99 @@ def convolve(input, weights, output=None, mode='reflect', cval=0.0, origin=0):
                                   True)
 
 
+def correlate1d(input, weights, axis=-1, output=None, mode="reflect", cval=0.0,
+                origin=0):
+    """One-dimensional correlate.
+    The array is correlated with the given kernel.
```
Comment (emcastillo) [inline cupyx/scipy/ndimage/filters.py:71]: One line break between summary and description 😄

File: cupyx/scipy/ndimage/filters.py [hunk 3]
```diff
@@ -64,6 +65,99 @@ def convolve(input, weights, output=None, mode='reflect', cval=0.0, origin=0):
...
+    """
+    weights, origins = _convert_1d_args(input.ndim, weights, origin, axis)
+    return _correlate_or_convolve(input, weights, output, mode, cval, origins,
+                                  False)
+
+
+def convolve1d(input, weights, axis=-1, output=None, mode="reflect", cval=0.0,
+               origin=0):
+    """One-dimensional convolution.
+    The array is convolved with the given kernel.
```
Comment (emcastillo) [inline cupyx/scipy/ndimage/filters.py:101]: ditto

File: cupyx/scipy/ndimage/filters.py [hunk 4]
```diff
@@ -64,6 +65,99 @@ def convolve(input, weights, output=None, mode='reflect', cval=0.0, origin=0):
...
+            origins[i] = -origins[i]
+            if wsize % 2 == 0:
+                origins[i] -= 1
+        origins = tuple(origins)
+    kernel = _get_correlate_kernel(mode, weights.shape, int_type,
+                                   origins, cval)
+    return _call_kernel(kernel, input, weights, output)
+
+
+@cupy.util.memoize()
```
Comment (emcastillo) [inline cupyx/scipy/ndimage/filters.py:150]: I think that memoization should be done per device

File: cupyx/scipy/ndimage/filters.py [hunk 5]
```diff
@@ -78,39 +172,105 @@ def _get_output(output, input, shape=None):
...
+#         axis += ndim
+#     if axis < 0 or axis >= ndim:
+#         raise ValueError('invalid axis')
+#     return axis
+
+
+def _convert_1d_args(ndim, weights, origin, axis):
+    if weights.ndim != 1 or weights.size < 1:
+        raise RuntimeError('incorrect filter size')
+    # axis = _check_axis(axis, ndim)
```
Comment (emcastillo) [inline cupyx/scipy/ndimage/filters.py:212]: pls, remove commented lines

File: cupyx/scipy/ndimage/filters.py [hunk 6]
```diff
@@ -148,6 +308,86 @@ def _generate_boundary_condition_ops(mode, ix, xsize):
...
+    value = '(*(X*)&data[{expr}])'.format(expr=expr)
+    if mode == 'constant':
+        value = '(({cond}) ? (X){cval} : {value})'.format(
+            cond=cond, cval=cval, value=value)
+    found = found.format(value=value)
+
+    operation = '''
+    {sizes}
+    {inds}
+    // don't use a CArray for indexing (faster to deal with indexing ourselves)
```
Comment (emcastillo) [inline cupyx/scipy/ndimage/filters.py:357]: I think this should be solved now in the master branch. it would be great if you can measure the performance difference with this and without this enabled.

File: cupyx/scipy/ndimage/filters.py [hunk 7]
```diff
@@ -205,7 +445,7 @@ def _generate_correlete_kernel(ndim, mode, cval, xshape, wshape, origin):
     return in_params, out_params, operation, name
 
 
-@util.memoize()
+@cupy.util.memoize()
```
Comment (emcastillo) [inline cupyx/scipy/ndimage/filters.py:448]: double check if per-device is needed please 😃

File: tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py [hunk 8]
```diff
@@ -121,6 +123,119 @@ def test_invalid_origin(self):
...
+        'cval': [0.0],
+        'origin': [0],
+        'adtype': [numpy.int32, numpy.float64],
+        'wdtype': [None],
+        'output': [None],
+        'filter': ['convolve1d', 'correlate1d']
+    })
+))
+@testing.gpu
+@testing.with_requires('scipy')
```
Comment (emcastillo) [inline tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py:169]: Should we specify a min/max version?

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by coderforlife** [comment]
If you read in the other PR, I am back, no longer working 16+ hour days... So we will need to discuss merging my most recent version with yours. I have fixed the `hasattr()` issue in my version along with merging the new minimum and maximum filters to use the same underlying CUDA code. I have not addressed the `CArray` but I know how to and I need to review other comments from this thread (just jumped onto this particular PR).

**Comment by mnicely** [comment]
Hey @coderforlife, glad to have you back! Please do whatever is easier :smile: 
You can merge or just copy my edits to yours. I'm not worried about any credit.

**Comment by coderforlife** [comment]
@emcastillo

I think these are the comments you brought up and things that were fixed in the @mnicely version:

1. Remove the section headers (like `##### Correlate and Convolve 1D #####`)
2. Add one line break summary and description in function docs
3. Memoization per-device
4. Switch out `_check_axis` for `cupy._util._normalize_axis_index`
5. Remove commented lines of code
6. Tests should be marked with a scipy version as appropriate

Things that haven't been fixed in @mnicely version:

1. `hasattr()` to `try/except`
2. Testing the speed with using CArray again now that it is smarter

I have addressed all of these at #3184 except the last one which I will get to later today. Additionally the one at #3184 uses a unified base for the new minimum/maximum filters added in 5da329e663126bfe8433bf3e89647dee329811e9 and the correlatlations/convolutions here (only one kernel generation function for all of them).

I also still have to add tests for the 1D min/max filters and a very other possible edge cases.

**Comment by emcastillo** [comment]
Thanks a lot for the follow up!
